crates/matrix-sdk/src/room/timeline/read_receipts.rs

```diff
@@ -0,0 +1,228 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{collections::HashMap, ops::Deref, sync::Arc};
+
+use eyeball_im::ObservableVector;
+use indexmap::IndexMap;
+use ruma::{
+    events::receipt::{Receipt, ReceiptEventContent, ReceiptThread, ReceiptType},
+    EventId, OwnedEventId, OwnedUserId, UserId,
+};
+use tracing::error;
+
+use super::{
+    inner::{RoomDataProvider, TimelineInnerState},
+    rfind_event_by_id, EventTimelineItem, TimelineItem,
+};
+
+struct FullReceipt<'a> {
+    event_id: &'a EventId,
+    user_id: &'a UserId,
+    receipt_type: ReceiptType,
+    receipt: &'a Receipt,
+}
+
+pub(super) fn handle_explicit_read_receipts(
+    receipt_event_content: ReceiptEventContent,
+    own_user_id: &UserId,
+    timeline_state: &mut TimelineInnerState,
+) {
+    for (event_id, receipt_types) in receipt_event_content.0 {
```

Comment:
Seems like we should implement `IntoIterator` for the event content types that are just newtypes over some collection.